Simplify API, esp. generic and async types

pursultani opened the pull request:
**"Java language designers will rue the day they added generics ... when they see our API"** :smile:

Commit:
refactor(api): Major interface refactoring

Including the following:

  - Reduced or simplified generics
  - Add enterypoint for API
  - Reduced or simplified async types
  - Renamed interfaces
  - Unified generic names

Commit:
feature(examples): Add HelloWorld example

Commit:
refactor(commons): Remove tagging interface for factory arguments

pursultani commented:
**WIP**

At the moment SPI build fails. I haven't refactored/re-implemented corresponding SPI interfaces.

mmnaseri commented on unchained-api/src/main/java/unchained/Selector.java:
```diff
@@ -0,0 +1,56 @@
+     * @param input
+     * @return
+     */
+    Hit<M> test(I input);
```
Rename to `select()`.

mmnaseri commented on unchained-api/src/main/java/unchained/Selector.java:
```diff
@@ -0,0 +1,56 @@
+
+    }
+
+    interface Hit<M> {
```
Rename to `Selection`

Commit:
feature(spi): Implement Configuration classes and factories

The old implementation of Lifecycle and Context classes are removed.
They must be adapted later to the new changes in the API.

Also, a workaround is added to run factory initialization at runtime.

Includes minor changes to the API.

Commit:
refactor(commons): Parametrize registerable keys

Commit:
fix(api): Better workaround for static initializer
